src/main/java/org/dataloader/DataLoaderHelper.java

```diff
@@ -296,49 +297,74 @@ private void possiblyClearCacheEntriesOnExceptions(List<K> keys) {
         We haven't been asked for this key yet. We want to do one of two things:
 
         1. Check if our cache store has it. If so:
-            a. Get the value from the cache store
-            b. Add a recovery case so we queue the load if fetching from cache store fails
+            a. Get the value from the cache store (this can take non-zero time)
+            b. Add a recovery case, so we queue the load if fetching from cache store fails
             c. Put that future in our futureCache to hit the early return next time
             d. Return the resilient future
         2. If not in value cache:
             a. queue or invoke the load
             b. Add a success handler to store the result in the cache store
             c. Return the result
          */
-        final CompletableFuture<V> future = new CompletableFuture<>();
+        final CompletableFuture<V> loadCallFuture = new CompletableFuture<>();
 
-        valueCache.get(cacheKey).whenComplete((cachedValue, getCallEx) -> {
-            if (getCallEx == null) {
-                future.complete(cachedValue);
+        CompletableFuture<V> cacheLookupCF = valueCache.get(cacheKey);
+        boolean cachedLookupCompletedImmediately = cacheLookupCF.isDone();
+
+        cacheLookupCF.whenComplete((cachedValue, cacheException) -> {
+            if (cacheException == null) {
+                loadCallFuture.complete(cachedValue);
             } else {
+                CompletableFuture<V> loaderCF;
                 synchronized (dataLoader) {
-                    queueOrInvokeLoader(key, loadContext, batchingEnabled)
-                            .whenComplete(setValueIntoCacheAndCompleteFuture(cacheKey, future));
+                    loaderCF = queueOrInvokeLoader(key, loadContext, batchingEnabled, loadCallFuture);
+                    loaderCF.whenComplete(setValueIntoValueCacheAndCompleteFuture(cacheKey, loadCallFuture));
+                }
+                //
+                // is possible that if the cache lookup step took some time to execute
+                // (e.g. an async network lookup to REDIS etc...) then it's possible that this
+                // load call has already returned and a dispatch call has been made, and hence this code
+                // is running after the dispatch was made - so we dispatch to catch up because
+                // it's likely to hang if we do not.  We might dispatch too early, but we will not
+                // hang because of an async cache lookup
+                //
+                if (!cachedLookupCompletedImmediately && !loaderCF.isDone()) {
+                    if (loaderOptions.getValueCacheOptions().isDispatchOnCacheMiss()) {
+                        dispatch();
+                    }
```

Comment:
This is the fix - this will dispatch if the cache get miss took time to complete before we enqueued the batch loader call.

The danger here is that it interferes with other batch calls on this loader (premature batch calling) but the alternative is that it will hang.

This balances that out

The other alternative is that people use `ScheduledDataLoaderRegistry` to periodically tick and dispatch on schedule